File: Backend/src/routes/groups.py
from src.controllers.groups import GroupController
from src.models.api.groups import Group
from fastapi import APIRouter, HTTPException
from typing import List, Dict, Any
import asyncio


router = APIRouter()


@router.post("/add_group")
async def add_group(group_name: Group) -> Dict[str, Any]:
    try:
        return await GroupController.add_group(group_name)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to add group: {str(e)}")


@router.post("/assign_device_to_group")
async def assign_device_to_group(device_mac: str, group_name: str) -> Dict[str, Any]:
    try:
        return await GroupController.assign_device_to_group(device_mac, group_name)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to assign device to group: {str(e)}")


@router.get("/get_all_groups")
async def get_all_groups() -> List[Dict[str, Any]]:
    try:
        return await GroupController.get_all_groups()
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to retrieve groups: {str(e)}")
    

@router.get("/one_group")
async def get_one_group(group_name: str) -> Dict[str, Any]:
    try:
        result = await GroupController.get_one_group(group_name)
        if not result:
            raise HTTPException(status_code=404, detail=f"Group not found: {group_name}")
        return result
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to retrieve group: {str(e)}")    
    

@router.delete("/delete_device_from_group")
async def delete_device_from_group(device_mac: str, group_name: str) -> Dict[str, Any]:
    try:
        return await GroupController.delete_device_from_group(device_mac, group_name)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to delete device from group: {str(e)}")    
    

@router.delete("/delete_group")    
async def delete_group(group_name: str) -> Dict[str, Any]:
    try:
        return await GroupController.delete_group(group_name)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to delete group: {str(e)}")

File: Frontend/src/api/groups.ts
const API_BASE_URL = 'http://localhost:8000';

export interface GroupSummary {
  group: string;
}

export interface GroupWithMembers extends GroupSummary {
  device_macs: string[];
}

interface GroupActionResponse {
  success?: boolean;
  message?: string;
  reason?: string;
}

export const fetchAllGroups = async (): Promise<GroupSummary[]> => {
  const response = await fetch(`${API_BASE_URL}/groups/get_all_groups`);
  if (!response.ok) {
    throw new Error('Unable to load groups');
  }

  const json = await response.json();
  if (!Array.isArray(json)) {
    return [];
  }

  return json as GroupSummary[];
};

export const fetchGroupWithMembers = async (
  groupName: string
): Promise<GroupWithMembers | null> => {
  const response = await fetch(
    `${API_BASE_URL}/groups/one_group?group_name=${encodeURIComponent(groupName)}`
  );

  if (!response.ok) {
    return null;
  }

  const json = await response.json();
  if (!json || typeof json !== 'object') {
    return null;
  }

  return {
    group: (json as any).group ?? groupName,
    device_macs: Array.isArray((json as any).device_macs)
      ? (json as any).device_macs
      : [],
  };
};

export const addGroup = async (payload: {
  group: string;
  device_macs?: string[];
}): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/groups/add_group`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const detail = await response.text();
    throw new Error(
      `Group could not be added (${response.status}): ${detail ?? ''}`
    );
  }
};

export const assignDeviceToGroup = async (
  deviceMac: string,
  groupName: string
): Promise<void> => {
  const response = await fetch(
    `${API_BASE_URL}/groups/assign_device_to_group?device_mac=${encodeURIComponent(
      deviceMac
    )}&group_name=${encodeURIComponent(groupName)}`,
    { method: 'POST' }
  );

  if (!response.ok) {
    const detail = await response.text();
    throw new Error(
      `Device could not be assigned (${response.status}): ${detail ?? ''}`
    );
  }
};

export const deleteDeviceFromGroup = async (
  deviceMac: string,
  groupName: string
): Promise<void> => {
  const response = await fetch(
    `${API_BASE_URL}/groups/delete_device_from_group?device_mac=${encodeURIComponent(
      deviceMac
    )}&group_name=${encodeURIComponent(groupName)}`,
    { method: 'DELETE' }
  );

  if (!response.ok) {
    const detail = await response.text();
    throw new Error(
      `Device could not be removed (${response.status}): ${detail ?? ''}`
    );
  }
};

// Placeholder for the upcoming endpoint; wired so the UI is ready once backend lands.
export const deleteGroup = async (groupName: string): Promise<GroupActionResponse> => {
  const response = await fetch(
    `${API_BASE_URL}/groups/delete_group?group_name=${encodeURIComponent(
      groupName
    )}`,
    { method: 'PUT' }
  );

  if (!response.ok) {
    const detail = await response.text();
    throw new Error(
      `Group could not be deleted (${response.status}): ${detail ?? ''}`
    );
  }

  const json = await response.json().catch(() => null);
  return (json ?? {}) as GroupActionResponse;
};
